Link all compatibility libraries when cross compiling and bootstrapping

edymtt opened the pull request:
Refactor the logic so to have a single target to reference the
compatibility libraries for the host, and use that when needed.

The main driver for this change is supporting the cross-compilation of
x86-64 on Apple Silicon.

Supports rdar://90307965

edymtt commented:
@swift-ci please test

edymtt commented:
@swift-ci please build toolchain

etcwilde reviewed (commented):
I think I would prefer having the list of libraries kept in a single place and tie them in with the `b0_deps` and `b1_deps` (https://github.com/apple/swift/blob/19831bea587afc9e2f8af07cda4c89ecbb90e4af/SwiftCompilerSources/CMakeLists.txt#L305-L311), that would help reduce the number of places I could possibly miss when I go and add a new compat lib.

compnerd commented on cmake/modules/AddSwift.cmake:
```diff
@@ -466,8 +466,10 @@ function(_add_swift_runtime_link_flags target relpath_to_lib_dir bootstrapping)
+      foreach(compat_library IN ITEMS "Concurrency" "DynamicReplacements" "50" "51")
+        set(conctarget "swiftCompatibility${compat_library}${vsuffix}")
+        target_link_libraries(${target} PUBLIC ${conctarget})
+      endforeach()
```
I think that rather than using `foreach` here, we should be explicitly spelling this out.  The build rules really should be generally declarative:

```
  target_link_libraries(${target} PUBLIC
    swiftCompatibilityConcurrency${vsuffix}
    swiftCompatibilityDynamicReplacements${vsuffix}
    swiftCompatibility50${vsuffix}
    swiftCompatibility51${vsuffix})
```

I think that hoisting this out a bit might be useful.  Perhaps we can get away with something like this instead?

```
add_library(swiftCompatibility INTERFACE)
target_link_libraries(swiftCompatibility INTERFACE
  swiftCompatibilityConcurrency${vsuffix}
  swiftCompatibilityDynamicReplacements${vsuffix}
  swiftCompatibility50${vsuffix}
  swiftCompatibility51${vsuffix})
```

This would mean that we define the set once and re-use it to avoid people forgetting a missing compatibility shim.

etcwilde replied:
So there used to be a `_link_built_compatibility_libs` function in `AddSwift` that would take a target and link all of the compatibility libraries against it and set the link directories, etc. That was removed here; https://github.com/apple/swift/pull/40370

The old definition was;
```
function(_link_built_compatibility_libs executable)
   set(platform ${SWIFT_SDK_${SWIFT_HOST_VARIANT_SDK}_LIB_SUBDIR})
   target_link_directories(${executable} PRIVATE
     ${SWIFTLIB_DIR}/${platform})
   add_dependencies(${executable}
     "swiftCompatibility50-${platform}"
     "swiftCompatibility51-${platform}"
     "swiftCompatibilityDynamicReplacements-${platform}")
 endfunction()
```

Bringing something like this back (maybe not setting this search path) so that there is only one place to update might be beneficial. Thoughts?

compnerd replied:
I think that the use of functions is a less desirable approach.  The interface library creates a placeholder name, which then propagates the dependencies and is more declarative rather than imperative for handling the dependencies (you just link that one library).  We can also inject any necessary flags (e.g. library search paths) through that for linking against the compatibility shims.

edymtt replied:
I took a stab in 02faa56a1a27b78a095e3fa397bddba9832af8d4 to incorporate this and Evan's feedback in the main discussion -- this will likely need some polishing (e.g. names and style), at this point I would like to check if this approach goes in the right direction.

edymtt replied:
Turns out I had to iterate more since -- see https://github.com/apple/swift/pull/60728#pullrequestreview-1085798384

edymtt commented on stdlib/toolchain/CMakeLists.txt:
```diff
@@ -48,8 +48,17 @@ list(APPEND CXX_COMPILE_FLAGS "-DSWIFT_COMPATIBILITY_LIBRARY=1")
+    add_subdirectory("Compatibility${compatibility_name}")
+  endforeach()
+
+  set(vsuffix "-${SWIFT_SDK_${SWIFT_HOST_VARIANT_SDK}_LIB_SUBDIR}-${SWIFT_HOST_VARIANT_ARCH}")
```
I preferred to keep the suffix here so target and architecture are explicit in the name of this target

edymtt commented:
@swift-ci please build toolchain

etcwilde commented on stdlib/toolchain/CMakeLists.txt:
```diff
@@ -48,8 +48,17 @@ list(APPEND CXX_COMPILE_FLAGS "-DSWIFT_COMPATIBILITY_LIBRARY=1")
-  add_subdirectory(Compatibility51)
-  add_subdirectory(CompatibilityDynamicReplacements)
-  add_subdirectory(CompatibilityConcurrency)
+  foreach(compatibility_name IN LISTS COMPATIBILITY_NAMES)
```
This section is way harder to read now.

I agree with @compnerd here. Something like the following is a bit more repetitive, though repetitions are kept close in the source, but it's way easier to read and know what's going on.
```
add_subdirectory(Compatibility50)
add_subdirectory(Compatibility51)
add_subdirectory(CompatibilityDynamicReplacements)
add_subdirectory(CompatibilityConcurrency)

add_library(HostCompatibilityLibs INTERFACE) 
target_link_libraries(HostCompatibilityLibs INTERFACE
  swiftCompatibilityConcurrency${vsuffix}
  swiftCompatibilityDynamicReplacements${vsuffix}
  swiftCompatibility50${vsuffix}
  swiftCompatibility51${vsuffix})
```
 
As long as the repetitions are kept close together in the code, it's manageable. The issue is having to remember to keep multiple files in sync with each other. Then those who need the compat libs can link against the `HostCompatibilityLibrary` and it's easy.

edymtt replied:
I admit in devising my initial change I thought I had to have a global variable (like in a general programming language) to reuse this in `SwiftCompilerSources/CMakeLists.txt` -- but in fact we can have that neatly defined in the proper place, and CMake will be able to find the target at generation time.

This has been reworked in 8b7d9152b7d -- let me know if that looks better.

edymtt committed:
Link all compatibility libraries when cross compiling and bootstrapping

Refactor the logic so to have a single target to reference the
compatibility libraries for the host, and use that when needed.

The main driver for this change is supporting the cross-compilation of
x86-64 on Apple Silicon.

Supports rdar://90307965

edymtt commented on SwiftCompilerSources/CMakeLists.txt:
```diff
@@ -301,12 +301,13 @@ else()
+      # because ultimately is used to specify a dependency for a
+      # custom target and, unlike `target_link_libraries`, such dependency
+      # would be lost at the generation of the build system.
+      get_property(compatibility_libs
```
On a build on Apple Silicon, this expands to

```
swift/SwiftCompilerSources/CMakeLists.txt(304):  get_property(compatibility_libs TARGET HostCompatibilityLibs PROPERTY INTERFACE_LINK_LIBRARIES )
swift/SwiftCompilerSources/CMakeLists.txt(307):  list(APPEND b0_deps swiftCompatibilityConcurrency-macosx-arm64;swiftCompatibilityDynamicReplacements-macosx-arm64;swiftCompatibility50-macosx-arm64;swiftCompatibility51-macosx-arm64 )
swift/SwiftCompilerSources/CMakeLists.txt(308):  list(APPEND b1_deps swiftCompatibilityConcurrency-macosx-arm64;swiftCompatibilityDynamicReplacements-macosx-arm64;swiftCompatibility50-macosx-arm64;swiftCompatibility51-macosx-arm64 )
```

edymtt commented:
@swift-ci please build toolchain

edymtt commented:
@swift-ci please test

edymtt commented:
@swift-ci please test macOS

edymtt commented:
@swift-ci please build toolchain macOS

edymtt commented:
@swift-ci please test macOS

edymtt commented:
@swift-ci please build toolchain Linux

edymtt commented:
@swift-ci please test Linux

compnerd reviewed (approved):
Thanks!  This definitely seems better.  Once the bootstrap path is cleaned up and uses CMake to build the Swift sources, it should allow us to avoid having to manually access the properties.

etcwilde reviewed (approved):
This is much better. LGTM. Thank you.
